scripts/generate_metadata.py

```diff
@@ -0,0 +1,172 @@
+            if os.path.isfile(file_path):
+                with open(file_path, "r", encoding="utf-8") as f:
+                    contents[filename] = f.read()
+    except Exception as e:
+        print(f"Error reading files in {folder_path}: {e}")
+    return contents
+
+
+def import_module_from_path(module_name: str, file_path: str) -> ModuleType:
+    spec = importlib.util.spec_from_file_location(module_name, file_path)
+    module = importlib.util.module_from_spec(spec)
+    spec.loader.exec_module(module)
+    return module
+
+
+def generate_tools_data() -> List[Dict[str, Any]]:
+    tools_data: List[Dict[str, Any]] = []
+    matches = find_folders_with_name()
+    for folder in matches:
+        print(f"\n Matched folder: {folder}")
+        subfolders = get_immediate_subfolders(folder)
+        for sub in subfolders:
+            print(f"  └── Subfolder: {sub}")
+            files = read_files_in_folder(sub)
+            tool_entry: Dict[str, Any] = {}
+            for fname, content in files.items():
+                if fname == INIT_PY:
+                    continue
+                if fname == COMPONENT_YAML:
+                    try:
+                        data = yaml.safe_load(content)
+                        tool_entry["author"] = data.get("author")
+                        tool_entry["tool_name"] = data.get("name")
+                        tool_entry["description"] = data.get("description")
+                    except Exception as e:
+                        print(f"Failed to parse YAML in {sub}: {e}")
+                        continue
+                else:
+                    file = Path(sub) / fname
+                    try:
+                        mod = import_module_from_path(fname, file)
+                        keys = ["ALLOWED_TOOLS", "AVAILABLE_TOOLS"]
+                        for k in keys:
+                            tools = getattr(mod, k, None)
+                            if isinstance(tools, list):
+                                tool_entry["allowed_tools"] = tools
+                                break
+                    except Exception as e:
+                        print(f"Failed to parse PY from {file}: {e}")
+                        continue
+
+            if tool_entry:
+                tools_data.append(tool_entry)
+
+    return tools_data
+
+
+def build_tools_metadata(tools_data: List[Dict[str, Any]]):
+    result = METADATA_TEMPLATE.copy()
+
+    for entry in tools_data:
+        author = entry.get("author", "")
+        tool_name = entry.get("tool_name", "")
+        allowed_tools = entry.get("allowed_tools", [])
+        if not allowed_tools:
+            print(
+                f"Warning: '{tool_name}' by '{author}' has no allowed tools/invalid format!"
+            )
+
+        for tool in entry.get("allowed_tools", []):
+            if tool not in result["tools"]:
+                result["tools"].append(tool)
+
+            result["toolMetadata"][tool] = {
```

Comment:
Shallow copy above means that the `toolMetadata` are also edited in the original constant.

Though I do not see why we need a copy and return the copied+edited value.